### PR TITLE
Use the correct source directory set on JVM

### PR DESCRIPTION
**Pull request by Goooler**
Same fixes borrowed from https://github.com/square/wire/pull/2310.

### CODE REVIEW

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt [hunk 1]
```diff
@@ -47,4 +44,8 @@
     }
 
     private fun FileCollection.existingFiles() = filter { it.exists() }
+
+    private val SourceSet.kotlinSourceSet: SourceDirectorySet
+        get() = ((this as HasConvention).convention.plugins["kotlin"] as? KotlinSourceSet)?.kotlin
+            ?: project.objects.sourceDirectorySet("empty", "Empty kotlin source set")
```
[the remark this answers is not in the document]
Comment (Goooler) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt]: 453338904bb67c57e54568fa3c4b051a5bf09543

**Comment by 3flex** [comment]
This logic doesn't make sense to me - it seems that in the case of a null source set, a detekt task would still be created, but `source` would be set to an empty value - so we'd be left with a task that will be skipped with a `NO-SOURCE` status.

If we need to filter out null source sets (and it would be good to understand what change has been made in the Kotlin Gradle Plugin that led to the change in behaviour) that should be done in a way that skips registration of a task that won't execute. I'd suggest logging a warning would be a good idea in that case as well.

**Review by 3flex — CHANGES_REQUESTED** [review]
This logic doesn't make sense to me - it seems that in the case of a null source set, a detekt task would still be created, but `source` would be set to an empty value - so we'd be left with a task that will be skipped with a `NO-SOURCE` status.

If we need to filter out null source sets (and it would be good to understand what change has been made in the Kotlin Gradle Plugin that led to the change in behaviour) that should be done in a way that skips registration of a task that won't execute. I'd suggest logging a warning would be a good idea in that case as well.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt [hunk 2]
```diff
@@ -2,15 +2,19 @@ package io.gitlab.arturbosch.detekt.internal
...
 import org.gradle.api.Project
 import org.gradle.api.file.FileCollection
+import org.gradle.api.file.SourceDirectorySet
 import org.gradle.api.internal.HasConvention
 import org.gradle.api.plugins.JavaPluginConvention
 import org.gradle.api.tasks.SourceSet
 import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
 
 internal class DetektJvm(private val project: Project) {
+    private val SourceSet.kotlin: SourceDirectorySet
```
Comment (cortinico) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt:14]: Is there a reason why this property is uppercase?

**Review by cortinico — COMMENTED** [review]
Thanks for sending this over 🙏

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt (hunk 2 above)
Comment (Goooler) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt:14]: It's lowercase.

Comment (cortinico) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt:14]: 🤦 Yeah my bad, it's an extension property

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt [hunk 3]
```diff
@@ -4,49 +4,55 @@ package io.gitlab.arturbosch.detekt.internal
...
-            source = kotlinSourceSet.kotlin
-            classpath.setFrom(sourceSet.compileClasspath.existingFiles(), sourceSet.output.classesDirs.existingFiles())
+    private fun Project.registerJvmDetektTask(
+        compilation: KotlinCompilation<KotlinCommonOptions>,
+        extension: DetektExtension,
+        inputSource: FileCollection
+    ) {
+        registerDetektTask(DetektPlugin.DETEKT_TASK_NAME + compilation.name.capitalize(), extension) {
+            setSource(inputSource)
+            classpath.setFrom(inputSource, compilation.compileDependencyFiles)
```
Comment (Goooler) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektJvm.kt:34]: Seems we can merge `DetektJvm` into `DetektMultiplatform`, see https://github.com/Goooler/detekt/pull/3.

**Comment by cortinico** [comment]
@3flex how do you feel about this change?

It looks promising on my end buy I'm lacking a bit of context. @Goooler could you expand on the PR description on why this is needed?

**Comment by 3flex** [comment]
I support this change, using the new approach - I'd still like to know what's changing in Kotlin 1.7.20 that means we have to make this change now, but putting that aside, we should do this anyway because:

1. The concept of Conventions is deprecated in Gradle, and KGP will drop them soon enough, requiring changes like this
2. As Goooler pointed out, it should allow consolidation of logic between MPP and Java variants at some point
3. Something like this will be needed for #4465 

I just wish we had better test coverage!

**Review by 3flex — APPROVED** [review]

**Comment by cortinico** [comment]
> The concept of Conventions is deprecated in Gradle, and KGP will drop them soon enough, requiring changes like this

TIL. Is this documented somehwere?

**Comment by 3flex** [comment]
This is probably most relevant:

https://docs.gradle.org/current/userguide/upgrading_version_7.html#all_convention_deprecation

**Comment by BraisGabin** [comment]
This is ready to merge, right? I would vote to merge it for 1.22.0-RC2.

**Comment by Goooler** [comment]
https://github.com/detekt/detekt/actions/runs/3196227698/jobs/5217876383

Seems `org.jetbrains.kotlin.gradle.plugin.mpp.pm20.targets` was removed from Kotlin 1.7.20, I'll fix it later.

**Pull request merged**